python/context.cpp

```diff
@@ -107,13 +118,20 @@ void register_contexts(pybind11::module& m) {
     pybind11::class_<proc_allocation_shim> proc_allocation(m, "proc_allocation",
         "Enumerates the computational resources on a node to be used for simulation.");
     proc_allocation
-        .def(pybind11::init<unsigned, pybind11::object>(),
-            "threads"_a=1, "gpu_id"_a=pybind11::none(),
-            "Construct an allocation with arguments:\n"
-            "  threads: The number of threads available locally for execution. Must be set to 1 at minimum. 1 by default.\n"
-            "  gpu_id:  The identifier of the GPU to use, None by default.\n")
+        .def(pybind11::init<unsigned, pybind11::object, bool, bool>(),
+             "threads"_a=1,
+             "gpu_id"_a=pybind11::none(),
+             "bind_procs"_a=false,
+             "bind_threads"_a=false,
+             "Construct an allocation with arguments:\n"
+             "  threads: The number of threads available locally for execution. Must be set to 1 at minimum. 1 by default.\n"
+             "  gpu_id:  The identifier of the GPU to use, None by default.\n")
         .def_property("threads", &proc_allocation_shim::get_num_threads, &proc_allocation_shim::set_num_threads,
             "The number of threads available locally for execution.")
+        .def_property("bind_procs", &proc_allocation_shim::get_bind_procs, &proc_allocation_shim::set_bind_procs,
+            "Try to bind MPI procs?")
```

Comment:
Specifically MPI proc? This is not what's in the API docs, which even advise against when using MPI.

Reply:
Read that again. The docs have -- as does the API -- _two_ levels of binding.
1. Proc: If managed externally, please don't try this, since it'll be redundant at best and likely detrimental.
2. Thread: See below.

So, if you want to go the full distance:
- Disable SLURM binding attempts (these are proc-level only!) `--cpu-bind=none`
- If using OpenMPI and your site is misconfigured, you might need to disable their's, too.
- Set `bind_procs` and `bind_threads` in `proc_allocation`.

Otherwise you'll want `procs=false`, `threads=true`.

Reply:
I can't retrace my steps. Docs are clear.